src/epik_eval/__main__.py

```diff
@@ -0,0 +1,483 @@
+
+        # Iterate minibatches
+        for i in range(0, batch.shape[0], args.minibatch_size):
+            with accelerator.accumulate(model):
+                # minimize padding
+                minibatch_max_len = seq_length[i: i + args.minibatch_size].max().item()
+                minibatch = batch[i: i + args.minibatch_size, :minibatch_max_len]  # minimizes padding
+                mini_attention_mask = attention_mask[i: i + args.minibatch_size, :minibatch_max_len]
+
+                if args.model_type == 'causal':
+                    output = model(input_ids=minibatch, attention_mask=mini_attention_mask)
+
+                    # causal language model must predict next word
+                    logits, mini_target = output.logits[:, :-1], minibatch[:, 1:]
+                    # (N, C), (N)
+                    logits, mini_target = logits.reshape(-1, logits.shape[-1]), mini_target.reshape(-1)
+                    loss = criterion(logits, mini_target)
+                elif args.model_type == 't5':
+                    mini_target_max_len = target_length[i: i + args.minibatch_size].max().item()
+                    mini_target = target[i: i + args.minibatch_size, :mini_target_max_len].contiguous()
+
+                    loss = model(input_ids=minibatch, attention_mask=mini_attention_mask, labels=mini_target).loss
+
+                epoch_loss += loss.item() * minibatch.shape[0]  # Undo the mean
+                epoch_samples += minibatch.shape[0]
+                accelerator.backward(loss)
+                optimizer.step()
+                lr_scheduler.step()
+                optimizer.zero_grad()
+
+                if args.model_type == 'causal':
+                    del loss, logits, output  # Free up memory
+                elif args.model_type == 't5':
+                    del loss
+
+    return epoch_loss / epoch_samples  # mean epoch loss
+
+
+def _evaluate(
+        accelerator,
+        model,
+        eval_dataloader,
+        eval_tokenizer,
+        device,
+        args,
+        epoch):
+
+    eval_dataloader.dataset.reset()
+    model.eval()
+
+    # Evaluation
+    results = []
+    while True:  # We have to iterate like this, because if not, the Accelerate dataloader merges samples from different batches, which are of different lengths and thus crashes.
+        prompt_id, prompt, attention_mask, nonduplicate_mask = next(iter(eval_dataloader))
+        
+        with torch.no_grad():
+            output = model.generate(
+                input_ids=prompt,
+                attention_mask=attention_mask,
+                max_new_tokens=args.answer_max_length,
+                pad_token_id=eval_tokenizer.pad_token_id,
+                synced_gpus=args.synced_gpus)
+        
+        results.append((prompt_id, output, nonduplicate_mask))
+
+        # Check if main process is finished iterating questions
+        if accelerator.gather(torch.tensor(eval_dataloader.dataset.finished, device=device)).any():
```

Comment:
Why `any()` should it not be `all()` or check only the main process?
Could you not check if `next(iter)` throws `StopIteration`?